rpcs3/Emu/System.cpp

```diff
@@ -494,8 +495,75 @@ bool Emulator::BootRsxCapture(const std::string& path)
 	return true;
 }
 
+void Emulator::LimitCacheSize()
+{
+	const std::string cache_location = Emulator::GetHdd1Dir() + "/cache";
+	if (!fs::is_dir(cache_location))
+	{
+		LOG_WARNING(GENERAL, "Cache does not exist (%s)", cache_location);
+		return;
+	}
+
+	const u64 size = fs::get_dir_size(cache_location);
+	const u64 max_size = static_cast<u64>(g_cfg.vfs.cache_max_size) * 1024 * 1024;
+
+	if (size <= max_size)
+	{
+		LOG_TRACE(GENERAL, "Cache size below limit: %d/%d", size, max_size);
+		return;
+	}
+
+	LOG_SUCCESS(GENERAL, "Cleaning disk cache...");
+	std::forward_list<std::tuple<std::string, u64, s64>> file_list{}; //name, size, last modified
+	fs::dir cache_dir{};
+	if (!cache_dir.open(cache_location))
+	{
+		LOG_ERROR(GENERAL, "Could not open cache directory");
+		return;
+	}
+
+	//retrieve items to delete
+	for (const auto &item : cache_dir)
+	{
+		if (item.name == "." || item.name == "..")
+			continue;
+
+		const std::string name = cache_location + "/" + item.name;
+		const u64 item_size = fs::is_dir(name) ? fs::get_dir_size(name) : item.size;
+		file_list.push_front(std::make_tuple(name, item_size, item.mtime));
+	}
+
+	//sort oldest first
+	file_list.sort([](auto left, auto right)
+	{
+		return std::get<2>(left) < std::get<2>(right);
+	});
+
+	//keep removing until cache is empty or enough bytes have been cleared
+	//cache is cleared down to 80% of limit to increase interval between clears
+	const u64 to_remove = static_cast<u64>(size - max_size * 0.8);
+	u64 removed = 0;
+	for (const auto &file : file_list)
+	{
+		const std::string &name = std::get<0>(file);
+		if (fs::is_dir(name))
+			fs::remove_all(name, true);
+		else
+			fs::remove_file(name);
+
+		removed += std::get<1>(file);
+		if (removed >= to_remove)
+			break;
+	}
+
+	LOG_SUCCESS(GENERAL, "Cleaned disk cache, removed %f MB", size / 1024.0 / 1024.0);
```

Comment:
Consider better formatting hint, such as %3.2f to avoid getting ridiculous IEEE representaions. Its not much use to get a fraction with 6 decimals, or worse, an exponent representation.